src/proxy/http/fallback.rs

```diff
@@ -0,0 +1,336 @@
+
+#[derive(Debug)]
+pub struct MakeFallback<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct Service<P, F, A> {
+    primary: P,
+    fallback: F,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct MakeFuture<P, F, A>
+where
+    P: Future,
+    F: Future,
+{
+    primary: Making<P>,
+    fallback: Making<F>,
+    _p: PhantomData<fn(A)>,
+}
+
+pub struct ResponseFuture<P, F, A>
+where
+    P: Future<Error = Error<A>>,
+    F: svc::Service<http::Request<A>>,
+{
+    fallback: F,
+    state: State<P, F::Future>,
+}
+
+pub enum Body<A, B> {
+    A(A),
+    B(B),
+}
+
+enum State<P, F> {
+    Primary(P),
+    Fallback(F),
+}
+
+enum Making<T: Future> {
+    NotReady(T),
+    Ready(T::Item),
+    Done,
+}
+
+pub fn layer<P, F, A>(primary: svc::Builder<P>, fallback: svc::Builder<F>) -> Layer<P, F, A> {
+    Layer {
+        primary,
+        fallback,
+        _p: PhantomData,
+    }
+}
+
+impl<P, F, A, M> svc::Layer<M> for Layer<P, F, A>
+where
+    P: svc::Layer<M> + Clone,
+    F: svc::Layer<M> + Clone,
+    M: Clone,
+{
+    type Service = MakeFallback<P::Service, F::Service, A>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        MakeFallback {
+            primary: self.primary.clone().service(inner.clone()),
+            fallback: self.fallback.clone().service(inner),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A, T> svc::Service<T> for MakeFallback<P, F, A>
+where
+    P: svc::Service<T>,
+    P::Error: Into<proxy::Error>,
+    F: svc::Service<T>,
+    F::Error: Into<proxy::Error>,
+    T: Clone,
+{
+    type Response = Service<P::Response, F::Response, A>;
+    type Error = proxy::Error;
+    type Future = MakeFuture<P::Future, F::Future, A>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        let primary_ready = self.primary.poll_ready().map_err(Into::into);
+        let fallback_ready = self.fallback.poll_ready().map_err(Into::into);
+        try_ready!(fallback_ready);
+        primary_ready
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        MakeFuture {
+            primary: Making::NotReady(self.primary.call(target.clone())),
+            fallback: Making::NotReady(self.fallback.call(target.clone())),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A> Clone for MakeFallback<P, F, A>
+where
+    P: Clone,
+    F: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            primary: self.primary.clone(),
+            fallback: self.fallback.clone(),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<P, F, A> Future for MakeFuture<P, F, A>
+where
+    P: Future,
+    P::Error: Into<proxy::Error>,
+    F: Future,
+    F::Error: Into<proxy::Error>,
+{
+    type Item = Service<P::Item, F::Item, A>;
+    type Error = proxy::Error;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let done =
+            self.primary.poll().map_err(Into::into)? && self.fallback.poll().map_err(Into::into)?;
```

Comment:
Is there something subtle going on here with `&&` and `Async::Ready(())` ?! This isn't clear to me... Would be clearer with `.is_ready()` or whatever.

Reply:
No, the "poll" fn on these returns `Result<bool, ...>`. Maybe it should be renamed?

Reply:
changed it to be a proper poll in 123b2b70f6fb082a0a2cdf4850e768fee8be4fc2, thanks! reads much better now IMO